Updates on How To Characterize Systems using GEOPM

lhlawson opened the pull request:
- Relates to #3175 from github issues
- Fixes #3175 change request from github issues.

A PR for covering validation and documentation of system characterization using GEOPM

lhlawson committed:
Started documenation on How To Characterize Systems with GEOPM

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Adding documentation of GPU F-E characterization

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

avilcheslopez commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -126,7 +126,7 @@ Without this, the uncore bandwidth characterization analysis scripts will not
+Additionally the ``test.sbatch`` or ``test.pbs`` should be modified to include the following
 experiment options, where the text within angle brackets (``<>``) needs to be
 replaced with relevant system (or administrator chosen) values::
 
```
Some notes for other details to update:
- The text below is still mainly based on Slurm, so it could be updated to also take PBS into account. Namely, there are references to `test.sbatch` and Slurm launch commands. Maybe we could use that nice box with a tab for each option (Slurm, PBS), as we do in our Getting Started guide (for Python, C, C++ code examples).
- In the note about extending the run time, add the line that needs to be updated (e.g. `#PBS -l walltime` for PBS).

lhlawson committed:
Provided info on how to check GPU-CA Fe fallbacks

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

avilcheslopez committed:
Include documentation for PBS in CPU activity agent page

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Minor CPU CA agent doc improvement

Add comments to InitControl file describing each setting.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

lhlawson committed:
First pass gpu activity agent test multi-vendor support

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

bgeltz commented on integration/test/test_gpu_activity_agent.py:
```diff
@@ -80,21 +81,42 @@ def launch_helper(experiment_type, experiment_args, app_conf, experiment_cli_arg
 
-        launch_helper(gpu_activity, experiment_args, parres.create_dgemm_appconf_cuda(mach, experiment_args), [])
+
+        path_nvidia = os.path.join(os.path.dirname(os.path.dirname(os.path.realpath(__file__))),
+                                                   "apps/parres/Kernels/Cxx11/dgemm-mpi-cublas")
+        path_intel = os.path.join(os.path.dirname(os.path.dirname(os.path.realpath(__file__))),
+                                                  "apps/parres/Kernels/Cxx11/dgemm-onemkl")
+        if os.path.exists(path_nvidia):
+            launch_helper(gpu_activity, experiment_args, parres.create_dgemm_appconf_cuda(mach, experiment_args), [])
+        elif os.path.exists(path_intel):
+            launch_helper(gpu_activity, experiment_args, parres.create_dgemm_appconf_oneapi(mach, experiment_args), [])
+        else:
+            self.fail("Neither NVIDIA or Intel dgemm variant was found")
```
```suggestion
        if util.get_service_config_value('enable_nvml') == '1':
            app_path = os.path.join(os.path.dirname(os.path.dirname(os.path.realpath(__file__))),
                                                    "apps/parres/Kernels/Cxx11/dgemm-mpi-cublas")
            app_conf = parres.create_dgemm_appconf_cuda(mach, experiment_args)
        elif util.get_service_config_value('enable_levelzero') == '1':
            app_path = os.path.join(os.path.dirname(os.path.dirname(os.path.realpath(__file__))),
                                                    "apps/parres/Kernels/Cxx11/dgemm-onemkl")
            app_conf = parres.create_dgemm_appconf_oneapi(mach, experiment_args), [])
        if not os.path.exists(app_path):
            self.fail("Neither NVIDIA or Intel dgemm variant was found")
        launch_helper(gpu_activity, experiment_args, app_conf, [])
```

Please do a similar refactor to the stream case below.

lhlawson committed:
Updated gpu-ca test to use get_service_config_value

Co-authored-by: Brad Geltz <brgeltz@gmail.com>

bgeltz commented on integration/test/test_gpu_activity_agent.py:
```diff
@@ -80,21 +81,42 @@ def launch_helper(experiment_type, experiment_args, app_conf, experiment_cli_arg
 
-    @util.skip_unless_nvml()
-    def test_gpu_activity_dgemm_nvidia(self):
+    @util.skip_unless_gpu()
+    @util.skip_unless_workload_exists("apps/parres/Kernels/Cxx11/")
```
```suggestion
```
This is unnecessary as it is specified at the top of the class already.

bgeltz commented on integration/test/test_gpu_activity_agent.py:
```diff
@@ -110,11 +132,12 @@ def test_gpu_activity_dgemm_nvidia(self):
 
-    @util.skip_unless_nvml()
-    def test_gpu_activity_stream_nvidia(self):
+    @util.skip_unless_gpu()
+    @util.skip_unless_workload_exists("apps/parres/Kernels/Cxx11/")
```
```suggestion
```
This is unnecessary as it is specified at the top of the class already.

lhlawson committed:
Apply suggestions from code review: Remove unnecessary skip decorators

Co-authored-by: Brad Geltz <brgeltz@gmail.com>

bgeltz committed:
Fixups for test_cpu_characterization and test_gpu_activity

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

WIP Use geopm-ctl-local with test_gpu_activity

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

Fix bad merge

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added GPU Freq Sweep and Characterization to GPU-CA test

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

bgeltz committed:
WIP Use geopm-ctl-local with test_gpu_activity

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

lhlawson committed:
WIP add test_gpu_characterization test and logic to gpu-ca agent test

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

avilcheslopez committed:
Add module for performing CPU CA characterization

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Refactor CPU CA tests to use CPU characterization module

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add module for performing GPU CA characterization

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add GPU characterization integration tests

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez commented on integration/test/test_gpu_activity_agent.py:
```diff
@@ -11,95 +11,99 @@
+                                       experiment_cli_args=experiment_cli_args)
+
+        base_dir = 'test_gpu_activity_output'
+        cls._gpu_ca_characterization = GPUCACharacterization(base_dir=base_dir)
```
Add comment explaining what's being done here.

avilcheslopez commented on integration/test/test_gpu_activity_agent.py:
```diff
@@ -11,95 +11,99 @@
-            cool_off_time=3,
-            enable_traces=False,
-            enable_profile_traces=False,
-            phi_list=None,
```
This should not have been removed.

avilcheslopez committed:
Enable process mode for minife

Also add minor improvements to CPUCharacterization to facilitate debug.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

lhlawson committed:
Fixed GPU Characterizatoin experiment config

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

avilcheslopez committed:
Get CPU-CA integration tests to pass

- Enable process mode and affinity for AIB and minife
- Don't LD_PRELOAD libgeopm
- Adjust problem size for CPU-CA characterization sweeps
- Misc. clean-up

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

lhlawson committed:
Temp fixes for unblocking frequency sweep experiments on system

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added report parsing updates for geopm-ctl-local runs

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added preload and program filter to parres

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added handling for geopm-ctl-local named reports

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
changed get_df to get_app_df to match recommendation script

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Update DGEMM GPU-CA test to remove < FOM check and increase iterations

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>
